include/picongpu/fields/laserProfiles/laserExpRampWithPrepulse.hpp

```diff
@@ -0,0 +1,189 @@
+        const float_X log2 = (t - t1) * math::log(a2);
+        return math::exp((log1 + log2)/(t2 - t1));
+    }
+
+    HDINLINE float_X
+    get_envelope(float_64 runTime)
+    {
+        float_X env = 0.0;
+        const bool before_preupramp = (-0.5 * RAMP_INIT * PULSE_LENGTH >
+            runTime);
+        const bool before_start = (runTime < 0.);
+        const bool before_peakpulse = (runTime < endUpramp);
+        const bool during_first_exp = (TIME_1 < runTime) and (runTime < TIME_2);
+        const bool after_peakpulse = (startDownramp <= runTime);
+        if (not before_preupramp and before_start)
+            env = AMP_1 * gauss(runTime - 0.) +
+                AMP_PREPULSE * gauss(runTime - TIME_PREPULSE);
+        else if (before_peakpulse)
+        {
+            const float_X ramp_when_peakpulse = extrapolate_expo(
+                TIME_2,
+                AMP_2,
+                TIME_3,
+                AMP_3,
+                endUpramp
+            ) / AMPLITUDE;
+            // if (ramp_when_peakpulse > 0.5) - I know, dead code :) didn't understand your comment about the other logging
+            //    throw std::invalid_argument("\n\nAttention, the intensities of the ramp are very large, the extrapolation to the time of the main pulse would give more than 50% of the pulse amplitude - this is not a gaussian pulse at all anymore, probably something wrong?!\n");
+
+            env += AMPLITUDE * (1.-ramp_when_peakpulse) * gauss(runTime-endUpramp);
+            env += AMP_PREPULSE * gauss(runTime - TIME_PREPULSE);
+            if (during_first_exp)
+                env += extrapolate_expo(
+                    TIME_1,
+                    AMP_1,
+                    TIME_2,
+                    AMP_2,
+                    runTime
+                );
+            else
+                env += extrapolate_expo(
+                    TIME_2,
+                    AMP_2,
+                    TIME_3,
+                    AMP_3,
+                    runTime
+                    );
+        }
+        else if (not after_peakpulse)
+            env = AMPLITUDE;
+    else // after startDownramp
+            env = AMPLITUDE * gauss(runTime-startDownramp);
+        return env;
+    }
+
+    HINLINE float3_X laserLongitudinal(uint32_t currentStep, float_X& phase)
+    {
+        float_X envelope;
+        float3_X elong(float3_X::create(0.0));
+
+        // a symmetric pulse will be initialized at position z=0 for
+        // a time of RAMP_INIT * PULSE_LENGTH + LASER_NOFOCUS_CONSTANT = INIT_TIME.
+        // we shift the complete pulse for the half of this time to start with
+        // the front of the laser pulse.
+
+        /* initialize the laser not in the first cell is equal to a negative shift
+         * in time
+         */
+        const float_64 runTime = (DELTA_T * currentStep - laserTimeShift -
+            0.5 * RAMP_INIT * PULSE_LENGTH);
+
+        const float_64 tau = PULSE_LENGTH * sqrt(2.0);
```

Comment:
Please use `math::sqrt()` this is also wrong in other laser implementations but only `math::*` will be platform independent.

Reply:
removed it, I don't use the tau ;)